File: crm-web/src/types/salesOrder.ts
export interface SalesOrderLine {
  id?: string;
  productId?: string;
  productCode?: string;
  productName: string;
  description?: string;
  quantity: number;
  unitPrice: number;
  unitOfMeasure?: string;
  discountPercent?: number;
  discountAmount?: number;
  taxRate?: number;
  taxAmount?: number;
  lineTotal?: number;
  sortOrder?: number;
}

export interface SalesOrder {
  id: string;
  orderNumber: string;
  quoteId?: string;
  account: {
    id: string;
    name: string;
  };
  contactId?: string;
  status: OrderStatus;
  orderDate: string;
  expectedDeliveryDate?: string;
  lines?: SalesOrderLine[];
  subtotal: number;
  discountAmount: number;
  taxAmount: number;
  total: number;
  currency: string;
  shippingStreet?: string;
  shippingCity?: string;
  shippingState?: string;
  shippingPostalCode?: string;
  shippingCountry?: string;
  notes?: string;
  assignedTo?: {
    id: string;
    fullName: string;
  };
  createdBy?: {
    id: string;
    fullName: string;
  };
  createdAt: string;
  updatedAt: string;
}

export interface SalesOrderListItem {
  id: string;
  orderNumber: string;
  status: OrderStatus;
  accountName: string;
  accountId: string;
  orderDate: string;
  expectedDeliveryDate?: string;
  total: number;
  currency: string;
  assignedToName?: string;
  createdByName?: string;
}

export interface SalesOrderLineRequest {
  productId?: string;
  productCode?: string;
  productName: string;
  description?: string;
  quantity: number;
  unitPrice: number;
  unitOfMeasure?: string;
  discountPercent?: number;
  taxRate?: number;
  sortOrder?: number;
}

export interface CreateSalesOrderRequest {
  accountId: string;
  contactId?: string;
  quoteId?: string;
  orderDate: string;
  expectedDeliveryDate?: string;
  lines?: SalesOrderLineRequest[];
  subtotal?: number;
  discountAmount?: number;
  taxAmount?: number;
  total?: number;
  currency?: string;
  shippingStreet?: string;
  shippingCity?: string;
  shippingState?: string;
  shippingPostalCode?: string;
  shippingCountry?: string;
  notes?: string;
  assignedToId?: string;
}

export interface UpdateSalesOrderRequest {
  status?: OrderStatus;
  contactId?: string;
  orderDate?: string;
  expectedDeliveryDate?: string;
  lines?: SalesOrderLineRequest[];
  subtotal?: number;
  discountAmount?: number;
  taxAmount?: number;
  total?: number;
  currency?: string;
  shippingStreet?: string;
  shippingCity?: string;
  shippingState?: string;
  shippingPostalCode?: string;
  shippingCountry?: string;
  notes?: string;
  assignedToId?: string;
}

export type OrderStatus =
  | 'DRAFT'
  | 'CONFIRMED'
  | 'SHIPPED'
  | 'DELIVERED'
  | 'CANCELLED';

export const ORDER_STATUS_CONFIG: Record<OrderStatus, { label: string; color: string }> = {
  DRAFT: { label: 'Brouillon', color: 'default' },
  CONFIRMED: { label: 'Confirmée', color: 'processing' },
  SHIPPED: { label: 'Expédiée', color: 'warning' },
  DELIVERED: { label: 'Livrée', color: 'success' },
  CANCELLED: { label: 'Annulée', color: 'error' },
};

File: crm-backend/crm-finance/src/main/java/com/opticrm/finance/entity/SalesOrder.java
package com.opticrm.finance.entity;

import com.opticrm.core.account.entity.Account;
import com.opticrm.security.entity.User;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.UuidGenerator;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "sales_orders")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class SalesOrder {

    @Id
    @UuidGenerator
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Column(name = "order_number", nullable = false, unique = true, length = 50)
    private String orderNumber;

    @Column(name = "quote_id")
    private UUID quoteId;

    // account_id OU chantier_id doit être renseigné (contrainte DB CHECK)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id", nullable = true)
    private Account account;

    // Référence UUID vers un Chantier (optionnel, alternatif à account)
    @Column(name = "chantier_id")
    private UUID chantierId;

    @Column(name = "contact_id")
    private UUID contactId;

    @Column(name = "status", nullable = false, length = 50)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private OrderStatus status = OrderStatus.DRAFT;

    @Column(name = "order_date", nullable = false)
    private LocalDate orderDate;

    @Column(name = "expected_delivery_date")
    private LocalDate expectedDeliveryDate;

    @Column(name = "subtotal", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal subtotal = BigDecimal.ZERO;

    @Column(name = "discount_amount", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(name = "tax_amount", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal taxAmount = BigDecimal.ZERO;

    @Column(name = "total", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal total = BigDecimal.ZERO;

    @Column(name = "currency", length = 3)
    @Builder.Default
    private String currency = "MAD";

    @Column(name = "shipping_street")
    private String shippingStreet;

    @Column(name = "shipping_city", length = 100)
    private String shippingCity;

    @Column(name = "shipping_state", length = 100)
    private String shippingState;

    @Column(name = "shipping_postal_code", length = 20)
    private String shippingPostalCode;

    @Column(name = "shipping_country", length = 100)
    private String shippingCountry;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @OneToMany(mappedBy = "salesOrder", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    @OrderBy("sortOrder ASC")
    private List<SalesOrderLine> lines = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to_id")
    private User assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id")
    private User createdBy;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private Instant updatedAt;

    public enum OrderStatus {
        DRAFT,
        CONFIRMED,
        SHIPPED,
        DELIVERED,
        CANCELLED
    }
}
